File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    vector<pair<int, int>> moves = {
        {-1, 0},
        {1, 0},
        {0, -1},
        {0, 1}
    };
    set<pair<int, int>> seen;

    bool dfs(string path, int x, int y, int idx, vector<vector<char>> board, string word) {
        if (path.length() > word.length()) {
            return false;
        }
        if (path == word) {
            return true;
        }

        bool flag = false;
        for (auto move: moves) {
            int dx = move.first, dy = move.second;
            int nx = x+dx, ny = y+dy;

            if (0 > nx or nx >= board.size() or 0 > ny or ny >= board[0].size()) {
                continue;
            }
            if (board[nx][ny] == word[idx] && !seen.count({nx, ny})) {
                seen.insert(make_pair(nx, ny));
                flag = dfs(path + board[nx][ny], nx, ny, idx+1, board, word);
                seen.erase(make_pair(nx, ny));
                if (flag) {
                    return true;
                }
            }
        }

        return flag;
    }

    bool exist(vector<vector<char>>& board, string word) {
        set<char> s;
        for (int i=0; i<board.size(); i++) {
            for (int j=0; j<board[0].size(); j++) {
                if (!s.count(board[i][j])) {
                    s.insert(board[i][j]);
                }
            }
        }

        for (auto c: word) {
            if (!s.count(c)) {
                return false;
            }
        }
        
        seen.clear();
        for (int i=0; i<board.size(); i++) {
            for (int j=0; j<board[0].size(); j++) {
                string path = "";

                if (board[i][j] == word[0]) {
                    seen.insert({i, j});
                    if (dfs(path + word[0], i, j, 1, board, word)) {
                        return true;
                    }
                    seen.clear();
                }
            }
        }

        return false;
    }
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        move = [(-1, 0), (0, 1), (1, 0), (0, -1)]

        def dfs(x, y, index, visited):
            visited[x][y] = True
            if board[x][y] != word[index]:
                return False
            
            if index + 1 == len(word):
                return True

            for dx, dy in move:
                nx, ny = x+dx, y+dy

                if 0 <= nx < m and 0 <= ny < n and not visited[nx][ny]:
                    if dfs(nx, ny, index+1, visited):
                        return True
                    visited[nx][ny] = False


        for x in range(m):
            for y in range(n):
                visited = [[False]*n for _ in range(m)]
                if dfs(x, y, 0, visited):
                    return True

        return False